src/runtime/pkg/device/drivers/utils.go

```diff
@@ -157,6 +159,98 @@ func checkIgnorePCIClass(pciClass string, deviceBDF string, bitmask uint64) (boo
 	return false, nil
 }
 
+func getMajorMinorFromDevPath(devPath string) (uint32, uint32, error) {
+	fi, err := os.Stat(devPath)
+	if err != nil {
+		return 0, 0, err
+	}
+
+	dev := fi.Sys().(*syscall.Stat_t)
+	return uint32(dev.Rdev >> 8), uint32(dev.Rdev & 0xff), nil
+}
+
+func extractIndex(devicePath string) (string, error) {
+
+	base := filepath.Base(devicePath)
+
+	const prefix = "vfio"
+	if !strings.HasPrefix(base, prefix) {
+		return "0", fmt.Errorf("unexpected device name format: %s", base)
+	}
+	return strings.TrimPrefix(base, prefix), nil
+}
+
+// Example: /sys/dev/char/506:0 -> ../../devices/pci0000:15/0000:15:01.0/0000:16:00.0/0000:17:02.0/0000:19:00.0/0000:1a:00.0/0000:1b:00.0/vfio-dev/vfio0
+func getBdfFromVFIODev(major uint32, minor uint32) (string, error) {
+	devPath := fmt.Sprintf("/sys/dev/char/%d:%d", major, minor)
+	realPath, err := filepath.EvalSymlinks(devPath)
+	if err != nil {
+		return "", fmt.Errorf("Failed to resolve symlink for %s: %v", devPath, err)
+	}
+	cleaned := filepath.Clean(realPath)
+	parts := strings.Split(cleaned, string(filepath.Separator))
+	if len(parts) < 3 {
+		return "", fmt.Errorf("path has fewer than 3 components: %s", cleaned)
+	}
+	// Check if the final component *resembles* a BDF by containing ':' and '.'
```

Comment:
seems redundant because this an ABI guarantee but probably OK.

here's what I had in my mind originally, btw: 
`filepath.Base(filepath.Dir(filepath.Join(realPath, "..")))`

Reply:
Remove you're right.